Feature/trigger in dataextensions

tiller1010 opened the pull request:
<!--
  Thanks for contributing, you're awesome! ⭐

  Please read https://docs.silverstripe.org/en/contributing/code/ if you haven't contributed to this project recently.
-->
## Description
Allows StaticPublishingTrigger to be added in DataExtensions

## Manual testing steps
- Add a DataExtension to a DataObject without StaticPublishingTrigger
- Also add the SiteTreePublishingEngine extension to that DataObject
- Confirm the instances of the DataObject have their URLs queued to be cached when saved/published

## Issues
Closes #211 

## Pull request checklist
<!--
  PLEASE check each of these to ensure you have done everything you need to do!
  If there's something in this list you need help with, please ask so that we can assist you.
-->
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [ ] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

tiller1010 committed:
improvements: check if record extensions do not have StaticPublishingTrigger interface

tiller1010 committed:
test: static trigger added in extension

GuySartorelli commented:
Can you please check all the boxes that apply in the PR checklist? That checklist is there so that you as the contributor can make sure you're doing things correctly before this gets reviewed. For example the commit message doesn't follow the guidelines.

ChloeSartorelli commented on src/Dev/DataExtensionAddsTrigger.php:
```diff
@@ -0,0 +1,27 @@
+use SilverStripe\StaticPublishQueue\Contract\StaticPublishingTrigger;
+use SilverStripe\StaticPublishQueue\Contract\StaticallyPublishable;
+
+class DataExtensionAddsTrigger extends DataExtension implements StaticallyPublishable, StaticPublishingTrigger, TestOnly
```
```suggestion
class DataExtensionAddsTrigger extends Extension implements StaticallyPublishable, StaticPublishingTrigger, TestOnly
```
`DataExtension` is deprecated, we shouldn't be using it.

ChloeSartorelli replied:
`TestOnly` classes should be in the `tests/` dir, _NOT_ `src/`

ChloeSartorelli commented on src/Dev/DataExtensionAddsTrigger.php:
```diff
@@ -0,0 +1,27 @@
+use SilverStripe\StaticPublishQueue\Contract\StaticPublishingTrigger;
+use SilverStripe\StaticPublishQueue\Contract\StaticallyPublishable;
+
+class DataExtensionAddsTrigger extends DataExtension implements StaticallyPublishable, StaticPublishingTrigger, TestOnly
```
Why do we add a new extension? The issue implies that the intention is for developers to be able to make their _own_ extension classes. Why do we need one in the module itself?

tiller1010 replied:
This implements the TestOnly interface. It was only for the added test.
Let me know if there is a better way to test with Extensions.

ChloeSartorelli replied:
That's fine, if this is only intended to be used in a test then it needs to be in the `tests/` dir as per https://github.com/silverstripe/silverstripe-staticpublishqueue/pull/214#discussion_r1917276773

ChloeSartorelli commented on src/Dev/DataObjectNoTrigger.php:
```diff
@@ -0,0 +1,15 @@
+use SilverStripe\Dev\TestOnly;
+use SilverStripe\ORM\DataObject;
+
+class DataObjectNoTrigger extends DataObject implements TestOnly
```
`TestOnly` classes should be in the `tests/` dir, _NOT_ `src/`

ChloeSartorelli commented on src/Extension/Engine/SiteTreePublishingEngine.php:
```diff
@@ -279,7 +279,10 @@ public function collectChanges($context)
 
             // The page does not include the required extension, and it doesn't implement a Trigger
-            if (!$siteTree->hasExtension(PublishableSiteTree::class) && !($siteTree instanceof StaticPublishingTrigger)) {
+            if (!$siteTree->hasExtension(PublishableSiteTree::class)
```
```suggestion
            if (
```
`PublishableSiteTree` implements `StaticPublishingTrigger` so there's no sense in checking for both.

ChloeSartorelli commented on tests/php/Extension/Engine/SiteTreePublishingEngineTest.yml:
```diff
@@ -22,3 +22,7 @@ SilverStripe\CMS\Model\SiteTree:
     Title: Page 6
     URLSegment: page-6
     ParentID: =>SilverStripe\CMS\Model\SiteTree.page5
+
```
```suggestion
```

ChloeSartorelli commented on tests/php/Extension/Engine/SiteTreePublishingEngineTest.yml:
```diff
@@ -22,3 +22,7 @@ SilverStripe\CMS\Model\SiteTree:
     ParentID: =>SilverStripe\CMS\Model\SiteTree.page5
+
+
+SilverStripe\StaticPublishQueue\Dev\DataObjectNoTrigger:
+  dataobject1:
```
If there's no data in it, there's no need to set up a fixture. Just `::create()->write()` in the test.

tiller1010 committed:
ENH Moved TestOnly classes to tests dir

- Replace DataExtension with Extension
- Used create instead of fixture entry in test
- Simplified engine extension check

GuySartorelli reviewed (approved):
LGTM, thanks for providing this improvement.
I'll fix up the commit message on merge.
